core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java

```diff
@@ -163,18 +163,56 @@ public void addConfigRoot(final PackageElement pkg, TypeElement clazz) {
             }
 
             final String fileName = computeExtensionDocFileName(member);
-            final List<ConfigDocItem> previousExtensionConfigDocKeys = foundExtensionConfigurationItems.get(fileName);
+            final List<ConfigDocItem> existingConfigDocItems = foundExtensionConfigurationItems.get(fileName);
 
-            if (previousExtensionConfigDocKeys == null) {
+            if (existingConfigDocItems == null) {
                 foundExtensionConfigurationItems.put(fileName, configDocItems);
             } else {
-                previousExtensionConfigDocKeys.addAll(configDocItems);
+                // merge two config doc items list
+                mergeConfigItemsIntoExistingOnes(existingConfigDocItems, configDocItems);
             }
         }
 
         return foundExtensionConfigurationItems;
     }
 
+    private void mergeConfigItemsIntoExistingOnes(List<ConfigDocItem> existingConfigDocKeys,
+            List<ConfigDocItem> configDocItems) {
+        for (ConfigDocItem configDocItem : configDocItems) {
+            if (configDocItem.isConfigKey()) {
+                existingConfigDocKeys.add(configDocItem);
+            } else {
+                ConfigDocSection configDocSection = configDocItem.getConfigDocSection();
+                boolean configSectionMerged = mergeSectionIntoPreviousExistingConfigItems(configDocSection,
+                        existingConfigDocKeys);
+                if (!configSectionMerged) {
+                    existingConfigDocKeys.add(configDocItem);
+                }
+            }
+        }
+    }
+
+    /**
+     * returns true if section is merged into one of the existing config items, false otherwise
+     */
+    private boolean mergeSectionIntoPreviousExistingConfigItems(ConfigDocSection section, List<ConfigDocItem> configDocItems) {
+        for (ConfigDocItem configDocItem : configDocItems) {
+            if (configDocItem.isConfigKey()) {
+                continue;
+            }
+
+            ConfigDocSection configDocSection = configDocItem.getConfigDocSection();
+            if (configDocSection.equals(section)) {
+                configDocSection.addConfigDocItems(section.getConfigDocItems());
+                return true;
+            } else if (mergeSectionIntoPreviousExistingConfigItems(section, configDocSection.getConfigDocItems())) {
```

Comment:
I asked before but you didn't answer my question: could you explain what you're doing here exactly? I don't understand why you need something recursive.

Reply:
Ah sorry, I think I missed your question or completely mis-understood it and ended up force pushing some re-naming I was finishing up.  

I opted for a recursive method to handle cases nested config section. This was we'll be able to merge a config section from two config roots, irrespective of their depth. e.g
```
class RunTimeConfigRoot {
  /**
   * foor-bar section
  */
  @ConfigDocSection
  BarConfigGroup group;
}
```
and 
```
class BuildTimeConfigRoot {
    FooConfigGroup group;
}

class FooConfigGroup {
   /**
    * foor-bar section
   */
   @ConfigDocSection
  DeepFooConfigGroup deepFoo;
} 
``` 

Is this something that we should deal with?